### PR TITLE
Fix collection install from source respects symlinks to directories

### LINKED ISSUE #78442: ansible galaxy install collection from git replaces directory symlinks with empty dir
### Summary

When a path in a collection is a symlink to a directory, the symlink is replaced with an empty dir instead of either a copy of the symlink target or the original symlink.

collection source:
```
$ tree -Ap --noreport ~/my_namespace/testcollection/collection/roles/php/templates/etc/php
[drwxr-xr-x]  roles/php/templates/etc/php
├── [drwxr-xr-x]  7.3
│   └── [drwxr-xr-x]  fpm
│       └── [drwxr-xr-x]  pool.d
│           └── [-rw-r--r--]  www.conf.j2
└── [lrwxrwxrwx]  7.4 -> 7.3
```

collection after installation (from git source):
```
$ tree -Ap --noreport ~/.ansible/collections/ansible_collections/my_namespace/testcollection/roles/php/templates/etc/php
[drwxr-xr-x]  roles/php/templates/etc/php
├── [drwxr-xr-x]  7.3
│   └── [drwxr-xr-x]  fpm
│       └── [drwxr-xr-x]  pool.d
│           └── [-rw-r--r--]  www.conf.j2
└── [drwxr-xr-x]  7.4
```

role `meta/requirements.yml` contains the following:
```
collections:
  - name: my_namespace.testcollection
    source: git+file:///home/twouters/my_namespace/testcollection#/collection/
    version: 2022.31.0-coll5
    type: git
```

contents of collection tar.gz after `ansible-galaxy collection build` seems to still be correct:
```
$ tar -tvf my_namespace-testcollection-2022.31.0-coll5.tar.gz | grep -F roles/php/templates
drwxr-xr-x 0/0               0 2022-08-03 11:31 roles/php/templates/etc/php/
lrw-r--r-- 0/0               0 1970-01-01 01:00 roles/php/templates/etc/php/7.4 -> 7.3
drwxr-xr-x 0/0               0 2022-08-03 11:31 roles/php/templates/etc/php/7.3/
drwxr-xr-x 0/0               0 2022-08-03 11:31 roles/php/templates/etc/php/7.3/fpm/
drwxr-xr-x 0/0               0 2022-08-03 11:31 roles/php/templates/etc/php/7.3/fpm/pool.d/
-rw-r--r-- 0/0             430 2022-08-03 11:31 roles/php/templates/etc/php/7.3/fpm/pool.d/www.conf.j2
```

installation from tar.gz with `ansible-galaxy collection install --force my_namespace-testcollection-2022.31.0-coll5.tar.gz` is fine:
```
$ tree -Ap --noreport ~/.ansible/collections/ansible_collections/my_namespace/testcollection/roles/php/templates/etc/php
[drwxr-xr-x]  roles/php/templates/etc/php
├── [drwxr-xr-x]  7.3
│   └── [drwxr-xr-x]  fpm
│       └── [drwxr-xr-x]  pool.d
│           └── [-rw-r--r--]  www.conf.j2
└── [lrwxrwxrwx]  7.4 -> 7.3
```

### Issue Type

Bug Report

### Component Name

ansible-galaxy

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.2]
  config file = /home/twouters/.ansible.cfg
  configured module search path = ['/home/twouters/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /home/twouters/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.5 (main, Aug  1 2022, 07:53:20) [GCC 12.1.0]
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
$
```

### OS / Environment

Arch Linux

### Steps to Reproduce

See summary

### Expected Results

See summary. I expect to either have a copy of the symlink target or preferably the original symlink.

### Actual Results

```console
See summary
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by sbettid**
##### SUMMARY
Fixes #78442 

The idea is to respect symlinks to directories when installing from source, by adding an extra check in the function used when installing from source/local repository, to avoid transforming symlinks to concrete directories. 

An integration test was added to monitor the mentioned use case.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME

ansible-galaxy-collection

##### ADDITIONAL INFORMATION

The problem was reproduced by using a test collection create as follows: 

`ansible-galaxy collection init my_namespace.test_collection`

Adding a folder inside the roles folder: `mkdir my_namespace/test_collection/roles/folderA`

Creating then a symlink to the created folder `ln -s folderA my_namespace/test_collection/roles/folderB`

Resulting roles folder: 

```
drwxr-xr-x. 1  20 28 set 09.51 folderA
lrwxrwxrwx. 1   7 28 set 09.00 folderB -> folderA
```

Result of installation from source, listing the directory `.ansible/collections/ansible_collections/my_namespace/test_collection/roles/`

Before

```
drwxr-xr-x. 1  20  2 ott 11.09 folderA
drwxr-xr-x. 1   0  2 ott 11.09 folderB
```
After

```
drwxr-xr-x. 1  20  2 ott 11.16 folderA
lrwxrwxrwx. 1   7  2 ott 11.16 folderB -> folderA
```

### CODE REVIEW

File: changelogs/fragments/78983-fix-collection-install-from-source-respects-dir-symlinks.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - ansible-galaxy-collection - respect symlinks to directories when installing from source or local repository (https://github.com/ansible/ansible/issues/78442)
```
Comment (s-hertel) [inline changelogs/fragments/78983-fix-collection-install-from-source-respects-dir-symlinks.yaml]: Fixes files too.

```suggestion
  - ansible-galaxy collection install - respect symlinks when installing from source or local repository (https://github.com/ansible/ansible/issues/78442)
```

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
@sbettid Thanks for fixing this!

**Comment by s-hertel** [comment]
@sbettid I'm second-guessing the directory part of this change, sorry. I think this breaks the intention on how symlink directories are supposed to work. From https://github.com/ansible/ansible/pull/69959#issuecomment-641653678:

> As for backwards compatibility with existing 2.9 releases here is what happens when install encounters a symlink in the collection artifact

> * File symlinks will create a copy of the symlink target
> * Dir symlinks will create an empty folder at the link path
>
> The latter is a breaking change on any new collections that are built with a directory symlink as instead of a copy being made of the folder and it's contents, an empty folder is there. This won't change any existing artifacts that are published but will only affect people running on ansible<=2.9.9 trying to install a new artifact that contains directory symbolic links. Considering the below points I think this is an ok breaking change to introduce

> * Ansible has never had any convention or scenario where it has used directory symbolic links, there shouldn't be any real reason to have a path to any plugin in any other folder
> * This keeps the behaviour of file and directory symbolic links similar to each other
> * Saves space on the collection as copies are no longer present
> * Because this change is planning on being backported to 2.9, ansible>=2.9.10 will be able to properly handle them, only older clients might miss out

**Comment by sbettid** [comment]
> @sbettid I'm second-guessing the directory part of this change, sorry. I think this breaks the intention on how symlink directories are supposed to work. From [#69959 (comment)](https://github.com/ansible/ansible/pull/69959#issuecomment-641653678):
> 
> > As for backwards compatibility with existing 2.9 releases here is what happens when install encounters a symlink in the collection artifact
> 
> > * File symlinks will create a copy of the symlink target
> > * Dir symlinks will create an empty folder at the link path
> > 
> > The latter is a breaking change on any new collections that are built with a directory symlink as instead of a copy being made of the folder and it's contents, an empty folder is there. This won't change any existing artifacts that are published but will only affect people running on ansible<=2.9.9 trying to install a new artifact that contains directory symbolic links. Considering the below points I think this is an ok breaking change to introduce
> 
> > * Ansible has never had any convention or scenario where it has used directory symbolic links, there shouldn't be any real reason to have a path to any plugin in any other folder
> > * This keeps the behaviour of file and directory symbolic links similar to each other
> > * Saves space on the collection as copies are no longer present
> > * Because this change is planning on being backported to 2.9, ansible>=2.9.10 will be able to properly handle them, only older clients might miss out

So this is breaking the initial intention on how symlinks are supposed to work during the collection installation, if I understood it correctly. How should this been addressed at this point?

**Comment by bcoca** [comment]
When building (what the code modified here does), symlinks from outside the collections should behave as before 'copy the contents', while symlinks within the collection should stay as symlinks.   

Note that when extracting symlinks to outside the collections should not be allowed, only symlinks to within it.

**Comment by sbettid** [comment]
@bcoca and @s-hertel  please forgive my ignorance and probably stupid questions.. From my understanding, the part on extracting symlinks to outside the collection should already covered in the `install_artifact` function, which has a check on this and which was not affected by this PR. Currently I think the missing part is the `copy the contents` when a symlink targets an element outside the collection when building, since with the current change it will stay a symlink what do you think? 

Excuse me again for bothering, I would only like to ensure the merged code on this does not lead to unwanted modifications.

**Comment by bcoca** [comment]
I have not reviewed the PR, I just added the comment to affirm the 'expected outcome' since that seemed to be in question above.

**Comment by s-hertel** [comment]
@sbettid Installing from source should end up with the same result as building + installing the artifact. `_build_collection_dir` doesn't make it obvious it's doing the same thing as `_build_collection_tar` and there's just a lack of test coverage for that. I'm not sure if there's an issue, but it seems like it's probably fine (after yet another visual review and some more manual testing). I merged it so if there ends up being an issue with it I will prioritize making sure it's fixed, no pressure on you.

**Pull request merged**